analyzetestgroups

```diff
@@ -0,0 +1,252 @@
+            continue
+        d = match.groupdict()
+        if p  == 'FIRST_LINE':
+            problemname = d['problemname']
+            if problemname != problempath.stem:
+                exit(f'FATAL: Problem directory does not match log file ({problemname}). Aborting...')
+            with open(problempath / 'problem.yaml', encoding = 'utf-8') as file:
+                problemtype = yaml.safe_load(file).get('type')
+                if problemtype != 'scoring':
+                    exit(f'FATAL: {problemname} is is not a scoring problem. Aborting...')
+            print (' '*80, end = '\r')
+            print(f'\033[01mScoring problem: {problemname}\033[0m')
+        elif p == 'START_SUBMISSION':
+            s = Submission(d['type'], d['filename'])
+        elif p == 'START_TESTGROUP':
+            tc_times: List[float] = []
+        elif p == 'AC_TC_RESULT':
+            print (problemname, s, end ='\r')
+            tc_times.append(float(d['time']))
+            tc_id = d['case']
+        elif p == 'TESTGROUP_GRADE':
+            grade = d['grade']
+            if d['type'] == 'sample': 
+                assert s is not None and len(s.verdicts) == 0
+            else:  
+                assert s is not None and len(s.verdicts) == int(d['number'])
+                assert grade != 'AC' or len(tc_times) # secret test group must have at least one tc for AC
+            s.verdicts.append(Verdict(grade, max(tc_times) if len(tc_times) else None))
+        elif p == 'END_SUBMISSION':
+            assert s is not None
+            s.points = int(d["points"] or "0")
+            s.maxtime = float(d["maxtime"])
+            submissions.append(s)
+        elif p == 'TIMELIMIT':
+            timelimits = int(d['limit']), int(d['safety'])
+        statusline = f'Submission {s}, test case {tc_id}'
+        print (' '*80, end = '\r')
+        print (statusline[:80], end = '\r')
+
+    assert timelimits is not None 
+    return timelimits, submissions
+
+def pretty_yn(value: bool) -> str:
+    # green y or red n
+     return '\033[32my\033[0m' if value else '\033[91mn\033[0m'
+
+def print_table(submissions):
+    alignto = max (len(str(sub)) for sub in submissions)
+    if alignto < len('Submission'):
+        alignto = len('Submission')
+    print("\033[01m", end = '')
+    print(f"{'Submission':{alignto}} Sample  ", end = ' ')
+    print(' '.join(f'Group {i} ' for i in range(1, num_secret_groups + 1)), end = ' ')
+    print('Pts Time  Expected\033[0m')
+    for s in submissions:
+        print (f"{s.filename:{alignto}}", end = ' ')
+        for verdict in s.verdicts:
+            print (f'{verdict:17}', end = ' ')
+        print(f"{s.points:3}", end = ' ')
+        print(f"{s.maxtime:4.2f}s", end = ' ')
+
+        # get the expected grades from the file (if exist), default to AC for AC submissions
+        expected_grades = None
+        path = problempath / 'submissions' / s.fullpath
+        with open(path, encoding = 'utf-8') as sourcefile:
```

Comment:
submissions can be directories (when there are multiple files, or when the main file needs to be named something specific), in which case you should search all files in the directory, probably recursively

Reply:
The difficulty is matching a source file name to the submission name. The full logic of `verifyproblem` is in `problemtools/problemtools/run/__init__.py`, returning a (list of) SourceCode objects, each of which has a `mainfile`. I feel disinclined to copy-paste all of that. It would be lovely if `verifyprobem -l info` reported the main file path, but no such luck. 

I’ll play around with it, but I fear it’ll be pretty hacky.

Reply:
Would it work to just have it search all files in the directory, instead of looking for the main one (which I agree is tricky)?

Reply:
Ah, ok. According to `Kattis/problemtools/problemtools/run/source.py`, 

   path (str): path of source code.  Can be either a single
                file or a directory (in which case the program is
                considered to consist of all files and subdirectories
                in the path).

And `self.name = os.path.basename(path)` . So I think I can infer the `name` from the output of `verifyproblem -l info`, and then just look for the expected grades in either the corresponding file (if it exists) or recursively go through everything in the single directory (which resides in `submissions/*/`), but not imitate the logic for finding `self.mainfile` in `source.py`.

Will do.

Reply:
Done in 8257719 .